java/src/main/java/ai/onnxruntime/OnnxSparseTensor.java

```diff
@@ -0,0 +1,920 @@
+      Buffer values,
+      int bufferPos,
+      long[] denseShape,
+      long[] valuesShape,
+      int onnxType)
+      throws OrtException;
+
+  /**
+   * Creates a sparse COO or block sparse tensor.
+   *
+   * <p>The buffers must be kept alive for the lifetime of the ORT sparse tensor object.
+   *
+   * @param apiHandle The ORT API pointer.
+   * @param allocatorHandle The allocator pointer.
+   * @param indicesData The indices.
+   * @param indicesBufferPos The indices position in bytes.
+   * @param indicesBufferSize The indices buffer size in longs.
+   * @param values The data.
+   * @param bufferPos The data position in bytes.
+   * @param denseShape The dense shape of the tensor.
+   * @param indicesShape The shape of the indices (a vector or matrix for COO, and a matrix for
+   *     block sparse).
+   * @param valuesShape The shape of the values (a vector for COO, and a block shape for block
+   *     sparse).
+   * @param onnxType The type of the values.
+   * @param sparsityType The sparsity type.
+   * @return A pointer to an ORT sparse tensor value.
+   * @throws OrtException If the tensor could not be created.
+   */
+  private static native long createSparseTensorFromBuffer(
+      long apiHandle,
+      long allocatorHandle,
+      Buffer indicesData,
+      int indicesBufferPos,
+      long indicesBufferSize,
+      Buffer values,
+      int bufferPos,
+      long[] denseShape,
+      long[] indicesShape,
+      long[] valuesShape,
+      int onnxType,
+      int sparsityType)
+      throws OrtException;
+
+  /**
+   * The type of the sparse tensor.
+   *
+   * <p>Should be synchronized with OrtSparseFormat in the C API.
+   */
+  public enum SparseTensorType {
+    /** Undefined sparse tensor. */
+    UNDEFINED(0),
+    /** COO sparse tensor. */
+    COO(1),
+    /** CSR or CSC sparse tensor. */
+    CSRC(2),
+    /** Block sparse tensor. */
+    BLOCK_SPARSE(4);
+
+    /** The int value mirroring OrtSparseFormat. */
+    public final int value;
+
+    private static final SparseTensorType[] values = new SparseTensorType[5];
+
+    static {
+      values[0] = UNDEFINED;
+      values[1] = COO;
+      values[2] = CSRC;
+      values[3] = UNDEFINED;
+      values[4] = BLOCK_SPARSE;
+    }
+
+    SparseTensorType(int value) {
+      this.value = value;
+    }
+
+    /**
+     * Maps from an int in native land into a SparseTensorType instance.
+     *
+     * @param value The value to lookup.
+     * @return The enum instance.
+     */
+    public static SparseTensorType mapFromInt(int value) {
+      if ((value > 0) && (value < values.length)) {
+        return values[value];
+      } else {
+        return UNDEFINED;
+      }
+    }
+  }
+
+  /**
+   * Abstract base class for Java sparse tensors
+   *
+   * <p>Will be sealed to {@link COOTensor}, {@link CSRCTensor} and {@link BlockSparseTensor} one
+   * day.
+   */
+  public abstract static class SparseTensor<T extends Buffer> {
+    private final long[] indicesShape;
```

Comment:
>] indicesShape; [](http://example.com/codeflow?start=23&length=15)

Would it not be better to have an array of shapes?

Reply:
We don't have a Java side shape class, and all the relevant methods expect a `long[]` as the shape.

Reply:
I would make a Index class with shape inside and all other relevant information. Then we would have no semantic overloading. SparseTensor would always contain what is needed. Alternatively, we could store Tensor classes in the indices array. Each tensor can have a shape and indices data. In C++ we simply project it to the buffer, here we can hold it.

Reply:
It's not accessible like that through the C API though, and that's what's represented here. My goal was to make something that was as transparent as possible, because there isn't a well accepted sparse tensor library in Java that we can bind to, so I figure people who want sparse behaviour will have created their own and be frustrated by any abstraction I try to layer in there.